### PR TITLE
fix(langchain): handle secret str api keys

### PR DESCRIPTION
**Pull request by albertjan**
Currently the anthropic chain implementation in langchain uses a pydantic SecretStr as an api key this is causing errors in our pipeline when ddtrace tries to format the api key. 

With this PR: https://github.com/langchain-ai/langchain/pull/12542 the OpenAI implementation will also start using a SecretStr. I'm sure at that point there will be a few more people asking why things are broken.

I'm struggling setting up and running the tests, riot doesn't print anything. And I have no experience with the cassettes testing methods. Can someone help with this? I think if we add a test that uses the Anthropic LLM we will see the failure before. And this will fix it.

I've updated the type comment to the function, but the env doesn't know about Pydantic so I don't know if this is a valid thing to do.

## Checklist

- [X] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [X] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [X] Change is maintainable (easy to change, telemetry, documentation).
- [X] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [X] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

**Comment by Yun-Kim** [comment]
Hi @albertjan, thanks for your contribution! I can take over testing and formatting and getting this merged.

**Comment by Yun-Kim** [comment]
The existing test cases should be sufficient since we continue to test latest versions of langchain - we'll be receiving `SecretStr` api keys (already are for AI21) once langchain continues to make these api key changes. Thanks again @albertjan for your contribution!

**Review by Yun-Kim — APPROVED** [review]

**Comment by albertjan** [comment]
You're welcome @Yun-Kim ☺️.

> The existing test cases should be sufficient since we continue to test latest versions of langchain 

A test using the `ChatAnthropic` would already exercise the `SecretStr` bit of the `_format_api_key` function.

**Review by majorgreys — COMMENTED** [review]
Do the existing tests cover the fixed case? I don't see any change to a span tag check.

**Comment by ZStriker19** [comment]
Agreeing with Tahir, can we add a test for  `span.set_tag_str(API_KEY, _format_api_key(api_key))  # override api_key for Pinecone` to verify the tag is changed as we expect?

**Comment by Yun-Kim** [comment]
@majorgreys @ZStriker19 the existing test cases cover this fixed case - langchain is in the process of switching over to use `SecretStr` api keys and already has been using `SecretStr` keys for the AI21 provider, which we already have a test case for. 

>  can we add a test for span.set_tag_str(API_KEY, _format_api_key(api_key))  # override api_key for Pinecone to verify the tag is changed as we expect?

That change is just a refactor, no functionality change as we're not getting that information through langchain.

File: ddtrace/contrib/langchain/patch.py [hunk 1]
```diff
@@ -695,7 +699,7 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
                 instance._index.configuration.server_variables.get("project_name", ""),
             )
             api_key = instance._index.configuration.api_key.get("ApiKeyAuth", "")
-            span.set_tag_str(API_KEY, "...%s" % api_key[-4:])  # override api_key for Pinecone
+            span.set_tag_str(API_KEY, _format_api_key(api_key))  # override api_key for Pinecone
```
Comment (Yun-Kim) [inline ddtrace/contrib/langchain/patch.py:702]: Note: this was a refactor, no functionality changes were made with this line

**Review by ZStriker19 — APPROVED** [review]

**Comment by albertjan** [comment]
Thanks all! 😄

**Pull request merged**